### ISSUE TITLE
Performance regression with canvas.drawPoints in stable releases after 2.10.5 

### ISSUE THREAD

**Issue by dfrtz** [flutter/flutter#104985]
**TLDR**: I believe there is a performance regression with `canvas.drawPoints`, in commit 02a182dd96 "Roll Engine from 605454e20143 to 0fd7713f2b5f" resulting in a 50%+ decrease in frame rendering speed in Flutter 3. It will take some additional time for me to collect the minimal repro, but I figured it was still worth while to bring to the attention of the community in case others encounter it while I work on it. Currently my solution/recommendation is to stay on 2.10.5, while also working on concurrently helping improve this in 3.0+.

## Details

**Target Platform:**
- Android

**Target OS version/browser:**
- Android 10/11/12

**Devices:**
- Pixel5a
- Pixel3 Emulator
- Other emulators

**Overview**
At first I thought I may be imagining it, but after I rolled back to 2.10.5 I confirmed an FPS improvement while idling on some heavier animated routes. The primary app I am working on has quite a bit of constant animations (it's a casual game), and I have tried to push the boundaries of what Flutter can handle with things like 2D particle effects, mock 3D perspectives, etc. A secondary concern I have is that it may be impacting other scenarios, they are just harder to notice whereas this one stands out easier. On 2.10.5 I can achieve ~60 FPS, whereas after upgrading to 3.0 the same pages often only manage ~30 FPS. It appears to be related to Skia itself based on the commit I narrowed it down to, but I am not 100% certain and appreciate any feedback.

I was curious if I could narrow down the impacted versions to when it first start, and I was able to isolate down to 2 tags to start:
```
2.11.0-0.1.pre # Good, ~60 fps
2.12.0-4.0.pre # Bad, ~30 fps
```

To attempt to narrow it down further, I checked out the 2.12 tag, and listed all the commits in between. This was ~800 commits, so I did a binary tree type checkout and performance test until I could narrow it down specifically to 02a182dd96:
```
# git checkout 2.12.0-4.0.pre
# git log --oneline 2.10.0-0.0.pre^..2.12.0-4.0.pre
commit 64a0138c937b1f7bd67e596571f4f33cd19c2b88 (HEAD, tag: 2.12.0-4.0.pre) - Bad perf.
...
02a182dd96 Roll Engine from 605454e20143 to 0fd7713f2b5f (3 revisions) (#98727) - First bad perf.
c50b02f2c3 Roll Engine from c5e89cbb7377 to 605454e20143 (3 revisions) (#98718) - Last good perf.
...
commit 95e555344dc746ad46ada9de9cc475042c63ec55 (tag: 2.10.0-0.0.pre) - Good perf.
```

I also tested the latest 3.1.0 tag on stable and master channels as well, but they behave the same. The last fast/safe flutter version details:
```
$ flutter --version
Flutter 2.11.0-0.0.pre.612 • channel unknown • unknown source
Framework • revision c50b02f2c3 (3 months ago) • 2022-02-17 22:44:23 -0500
Engine • revision 605454e201
Tools • Dart 2.17.0 (build 2.17.0-118.0.dev) • DevTools 2.10.0
```

First slow flutter version details:
```
$ flutter --version
Flutter 2.11.0-0.0.pre.613 • channel unknown • unknown source
Framework • revision 02a182dd96 (3 months ago) • 2022-02-18 02:24:27 -0500
Engine • revision 0fd7713f2b
Tools • Dart 2.17.0 (build 2.17.0-119.0.dev) • DevTools 2.11.0
```

This commit appears to be minor changes, but based on the PR there were 3 engine changes in this commit spanning Dart, Skia, and Flutter:
https://github.com/flutter/flutter/pull/98727
```
2022-02-18 skia-flutter-autoroll@skia.org Roll Dart SDK from 6894cf5d241c to 9ac87a1a1ca8 (1 revision) (flutter/engine#31534)
2022-02-18 gw280@google.com Add a complexity scoring class for Metal and OpenGL (flutter/engine#31417)
2022-02-18 skia-flutter-autoroll@skia.org Roll Skia from b5282fea2312 to 418fdebf1cca (1 revision) (flutter/engine#31531)
```

If I had to guess, it is probably related to the Skia rollup or the scoring changes? But I looked at 418fdebf1cca in Skia, and 9ac87a1a1ca8 in Dart, and I didn't see anything that stood out. I also checked my third party dependencies for any changes, and they showed the exact same libs in the pubspec.lock when switching between the fast/slow commits.

I have used the CPU profiler to narrow down some performance issues in the past, so I did a 10 second capture on these pages on the good and bad commit for comparison. The graphs look almost identical, leading me to further believe the issue may be in Skia. I checked for comparisons to flutter 2.10.5 and 2.5.0, the previous versions I was running, and while 2.5 looks drastically different in the flame charts, 2.10 seems to work great and looks similar.

Here are screen caps of the "stable" perfs when I remove `drawPoint` references, and keep all other animations going. The perf appears consistent between the 2 versions at a basic level:
<details>
<summary>Perf Timeline Snapshots Without drawPoint Calls (Stable perf with other animations)</summary>
Good c50b02f2c3 commit:
<img width="969" alt="c50b02_without_points" src="https://user-images.githubusercontent.com/29335974/171052830-5f516fdb-001f-47a7-939d-932b25f9a48d.png">
Problematic 02a182dd96 commit:
<img width="963" alt="02a182_without_points" src="https://user-images.githubusercontent.com/29335974/171052874-dd2d2ef6-55b9-4a8c-a989-6b7aecc02b10.png">
</details>

Here are with `drawPoint` calls included. There is some overhead from the profiling/tracking turned on, since I can hit 60 FPS on the good version without it, but even so there is about a 2x increase (16 vs 32 ms) in the raster time:
<details>
<summary>Perf Timeline Snapshots With drawPoint Calls (2x regression)</summary>
Good c50b02f2c3 commit:
<img width="971" alt="c50b02_with_points" src="https://user-images.githubusercontent.com/29335974/171053406-53d6af54-97eb-4b80-b2f9-dcf238cfdf92.png">
Problematic 02a182dd96 commit:
<img width="960" alt="02a182_with_points" src="https://user-images.githubusercontent.com/29335974/171053416-efe30bff-e159-4ce9-8209-2ee7497ca040.png">
</details>

And to push the boundaries further, here is a comparison of what the "maximum strength" of these `drawPoints` calls could theoretically be in one of my app routes that are currently smooth. There is some overhead from the profiling/tracking turned on, since I can hit 60 FPS on the good version without it, but there is about 4-5x increase (32 vs 139 ms) in raster time:
<details>
<summary>Perf Timeline Snapshots With "max" drawPoint Calls (4-5x regression)</summary>
Good c50b02f2c3 commit:
<img width="967" alt="c50b02_with_points_max" src="https://user-images.githubusercontent.com/29335974/171053543-f987d7c7-dc0b-40d5-a2a5-5800734e80c4.png">
Problematic 02a182dd96 commit:
<img width="968" alt="02a182_with_points_max" src="https://user-images.githubusercontent.com/29335974/171053523-b210544c-1d23-484e-803a-9e01af92918c.png">
</details>

I have tested debug/profile/release modes, and they all behave the same way. This behavior is noticeable on both my physical Pixel 5a and Android emulators. For example, debug mode on 2.10.5 runs smoother on these animated pages than release mode on 3.0.

I was hoping someone may have encountered this before me, but I do not believe so after looking over open perf related issues and based on the timeframe around when it happened (likely in last 2 months based on the commit). I am working on a minimum viable repro to fill in the blanks, but I wanted to see if there were any initial thoughts about the commit I isolated it too, or things to try. I understand if there is not much feedback that can be provided until I can come up with a "minimal" app, so I am working on isolating it further and providing additional details. However, I wanted to try to bring some attention to it while I try to narrow it down even further, in case someone else hits similar issues and/or may be able to isolate a minimum repro before I can.

I appreciate all the hard work that has gone in to Flutter, it's great stuff! I like to push boundaries and make things even more awesome 😄 

## Logs

<details>
<summary>Logs</summary>

Pending minimal repro in progress.
```
# Pending `flutter analyze` and additional logs from minimal repro in progress.
```
</details>

<details>
<summary>Doctor Output</summary>

Details about my dev setup when on the first slow commit (02a182dd96). I am not working on mac/ios versions currently, even though I plan to at some point, so I know my xcode is not set up and I havent tested if iOS is impacted.
```
$ flutter doctor -v
[✓] Flutter (Channel unknown, 2.11.0-0.0.pre.613, on macOS 11.6 20G165 darwin-x64, locale en-US)
    • Flutter version 2.11.0-0.0.pre.613 at /Users/dfritz/Library/flutter
    • Upstream repository unknown
    • Framework revision 02a182dd96 (3 months ago), 2022-02-18 02:24:27 -0500
    • Engine revision 0fd7713f2b
    • Dart version 2.17.0 (build 2.17.0-119.0.dev)
    • DevTools version 2.11.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/dfritz/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • Java binary at: /Users/dfritz/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/212.5712.43.2112.8609683/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✗] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
        sudo xcodebuild -runFirstLaunch
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage
        on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Users/dfritz/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/212.5712.43.2112.8609683/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] Android Studio (version 2020.3)
    • Android Studio at /Users/dfritz/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7678000/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] Connected device (2 available)
    • Pixel 5a (mobile) • 17311JECB09831 • android-arm64  • Android 12 (API 32)
    • Chrome (web)      • chrome         • web-javascript • Google Chrome 102.0.5005.61

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```
</details>

**Comment by jonahwilliams** [comment]
It sounds like the likely culprit is "2022-02-18 gw280@google.com Add a complexity scoring class for Metal and OpenGL (flutter/engine#31417)", because that would change the threshold at which a picture would qualify for raster caching. If your example was previously above the threshold and is now below it, that would explain some performance difference.

That said, the complexity scores are intended to be a way to gauge whether or not it is worthwhile to raster cache a picture. if there is a significant regression in performance due to a picture no longer being raster cached than it may be that there is an edge case in the complexity scoring that we've missed.

To help us determine the cause, a minimal reproduction of the performance regression would be required.

**Comment by dfrtz** [comment]
Thanks for the tips/hints! I was just looking at (flutter/engine#31417) further and I do agree it is likely the cause.

The difficult part I am finding is that I just finished up another attempt at trying to create a minimal repro from the basic steps I provided (which are still a WIP), but I was not able to trigger it. Even with scaling between 1,000-20,0000 points being drawn, I can't seem to trigger it with basic calls. I did hit a major up front cost to render it on the first paint, but it appears that the ongoing cost is not being incurred. Interestingly this is opposite to what I am seeing my full app, where there is little up front cost, but large ongoing costs.

I will keep working on the minimal repro, there is obviously some unique combination with these game pieces I have that triggers it. I will have to keep looking at the complexity changes to see what about my drawPoints calls is possibly triggering the threshold adjustments. Maybe something is hitting certain AA or hairline values my simple repro attempts are not.

**Comment by dfrtz** [comment]
I haven't had a chance to look over the weighting changes in more details to help try to narrow it down further (which it's a little out of my wheelhouse so I may not be able to anyways), but after a few attempts I have been able to create a basic repro. It may look like a funny ms paint job when run, but it does the trick while simulating the scenario of "game pieces" that may be removed/added over time.

Summary of attempts to repro:
- Few render widgets (<5) with large amounts of drawPoints calls: No. But there is a pause on first render depending on how many "pieces" there are.
- Many render widgets (~100) with few drawPoints calls: No.
- Extreme render widgets (~500) with few drawPoints calls: Not really a repro, its slow on both commits. Considering not related.
- Medium render widgets (~50-60) with medium amount of drawPoints calls: Bingo.
    - Code available at: https://gist.github.com/dfrtz/216dda1d84591e4745da3db3a76ee45d

I am sure that if the "many/extreme widgets" amounts combined with more than just 1-2 drawPoints could also repro, but the medium/medium was easiest to show the contrast in performance between c50b02f2c3 and 02a182dd96 without resulting in them both being slow. Here are some snaps to show the behavior from the repro code.

<details>
<summary>Perf comparison with 64 objects (10ms vs 40ms)</summary>
64 objects, speedy c50b02f2c3 commit:
<img width="930" alt="c50b02f2c3_repro_64_objects" src="https://user-images.githubusercontent.com/29335974/171751898-3ecad0c4-2821-4a63-8c96-ac65c17ec1d6.png">

64 objects, problematic 02a182dd96 commit:
<img width="965" alt="02a182dd96_repro_64_objects" src="https://user-images.githubusercontent.com/29335974/171751910-69ebe6aa-cc3b-45f4-b36b-5e6d0235dbe5.png">
</details>

I was curious just how far I could push it on the older commit, and even with double the objects it performs about 3-4x faster. Its not until 4x the objects that it starts to breach below 60 FPS, but even then it is still rendering faster than 64 on the new commit.
<details>
<summary>Perf with with 128 and 256 objects on c50b02f2c3</summary>
128 objects, still very capable of 60 FPS:
<img width="943" alt="c50b02f2c3_repro_128_objects" src="https://user-images.githubusercontent.com/29335974/171751924-92186728-29e6-43fc-af07-248c762faeb2.png">

256 objects, now redlining but still half as high as 64 objects on 02a182dd96:
<img width="935" alt="c50b02f2c3_repro_256_objects" src="https://user-images.githubusercontent.com/29335974/171751929-1f90f257-21ed-49bc-8158-7a06cfc25568.png">
</details>

It appears the primary trigger is to keep the individual drawings in LeafRenderObjectWidgets simple/medium complexity, but have many of them. For comparison I also tried grouping many of the "pieces" into a single widget. I did not hit ongoing frame render costs, but experienced large render costs/freezes up front. While there may be some balance to strike with grouping if wanting to push the boundaries higher than 128-256 objects, lockups have a more negative impact than short slowness for the same timeframe.

Additional things to note about this repro compared to my prod scenario:
- In my full app it never starts off 64-128+ "pieces" at once (probably about 16-32), but it can hit this amount over time.
- My prod pieces are actually a bit more complex than these simple "cactuar like" figures, but they were the easiest way to demonstrate the regression.
- When grouping pieces together in a single widget, there ends up being a large cost/freeze, and then it is smooth. I assume this is because it exceeded the new complexity values.
- When kept separate, there is no large cost/freeze upfront, but it stays choppy the entire time on 02a182dd96.
    - Paint is only called once, but it appears raster caching is no longer kicking in.
    - This is silky smooth on c50b02f2c3 even with twice as many objects.
    - It is not until 4x the amount of objects c50b02f2c3 starts to break below 60 FPS, but even then it is half as costly as 1/4 the objects on 02a182dd96.

Perhaps my use case is an edge case or an originally unintended use case. I know that there is a push with 3.0 for casual games, but most of the push is through Flame, whereas mine is a custom solution. I am using leaf renderers and have them as repaint boundaries, but that is because they have had great performance. While I did plan to combine some of my prod pieces into larger groups, it appears that optimizing in this way may have a different side effect with the long pauses.

In my mind the ideal scenario is to be able to reach the same level of performance as before, while using separate objects/widgets, because that feels the most flexible from a development standpoint. It also seems to provide the best experience for this kind of use case, where the objects are "complex" but swap in/out often. Grouping might be ideal in other scenarios where they never change, but if a single piece is removed here, then a grouping would likely incur the high cost again.

**Comment by jonahwilliams** [comment]
Thank you for the additional details and the reproduction of the issue. I'll take a look at it when I get a chance

Though they're not well documented, there are a few techniques you can use to claw back some performance:

1. If your Picture is fairly stable, rather than using a repaint boundary, instead consider creating your own PictureRecorder/Canvas to imperatively construct a picture, then use Picture.toImage and draw the resulting image without a repaint boundary.

2. You could set https://api.flutter.dev/flutter/widgets/CustomPaint/isComplex.html to true, which overrides the raster cache behavior . This is easier to do on a custom painter.

**Comment by dfrtz** [comment]
Thanks for the extra tips! I did try the isComplex option in some prototypes, but unfortunately that was not working as I had hoped and led to my usage of leaf renderers for the performance. I haven't had a chance to try a PictureRecorder version yet, however it was on my list of potential future enhancements, so I'll have to give that a go and see how it turns out.

**Comment by dfrtz** [comment]
I gave a PictureRecorder version of these widgets a try, and the results were very positive for the static versions. I believe I may have found another impact from the engine changes in 02a182dd96 however, since not all widgets provide as much control as CustomPainters.

For example, with the minimal repro code I was able to push 4x+ the amount of objects compared to the LeafRenderer versions on both versions. Compared to c50b02f2c3 I could push 4x+, and compared to 02a182dd96 I could push at least 8x+ for the same cost. It was a little tricky due to the pixel density scaling needed on both the image creator and painter, but this is a short snippet of the generator.
```
Future<ui.Image> toImage() async {
  ui.PictureRecorder recorder = ui.PictureRecorder();
  Canvas canvas = Canvas(recorder);
  double pixelRatio = ui.window.devicePixelRatio;
  canvas.scale(pixelRatio, pixelRatio);
  paint(canvas, Offset(radius, radius));
  int imageSize = (radius * 3 * pixelRatio).ceil();
  return await recorder.endRecording().toImage(imageSize, imageSize);
}
```

And here is a comparison on 02a182dd96 of 64 objects via renderers vs 1024 objects via images. About a 16x increase for the same cost.
<details>
<summary>02a182dd96 64 leaf renderers vs 1024 image painters</summary>
64 renderers, 02a182dd96:
<img width="930" alt="c50b02f2c3_repro_64_objects" src="https://user-images.githubusercontent.com/29335974/171751898-3ecad0c4-2821-4a63-8c96-ac65c17ec1d6.png">

1024 image painters, 02a182dd96:
<img width="942" alt="02a182dd96_picturerecorder_1024objs" src="https://user-images.githubusercontent.com/29335974/172062698-6052ff1c-3801-4241-9d7f-1a0ff34fb3e3.png">
</details>

As mentioned it is more performant than c50b02f2c3 with 256 objects as well, just not quite as drastic of a comparison. This implementation works best for the static/non-animated versions, so it is worth it to make the switch for those. I just had to breakout the painting functions so it can be reused across the custom painter and the renderboxes. The renderer version still works best for the animated instances, but those are a smaller subset. A more interesting detail from this experiment I have discovered is that even with this change, my overall performance on my prod app is still a bit lower on 3.0+ than when running on 2.10.5. The PictureRecorder version of these pieces does help push the boundaries even further, which is awesome, but it seems 2.10.5 is still the sweet spot for performance.

To elaborate further on the potential additional impact, it is two part: you need to force isComplex on the painters, and widgets without this option may be impacted. Switching to PictureRecorder for painting the images exposed the need for an additional consideration in >= 02a182dd96 involving non-drawPoint operations with CustomPainters. On c50b02f2c3 isComplex did not have to be forced to true, but on 02a182dd96 it does, otherwise a severe performance drop will be encountered. As part of this investigation I noticed that if I move from a custom shadow widget (with a duplicate set of image painters), to painting the images with an ImageFilter directly, I could squeeze out even more performance on c50b02f2c3. The cost goes from <10ms to >60ms for this use case if forgetting to force isComplex.

Code snippets indicating how this was changed to use the PictureRecorder images to further optimize on c50b02f2c3:
```
final Paint _shadowPainter = Paint()
  ..color = Colors.black.withOpacity(0.66)
  ..imageFilter = ui.ImageFilter.blur(sigmaX: 6, sigmaY: 6)
  ..colorFilter = ColorFilter.mode(Colors.black, BlendMode.srcATop);

canvas.drawImage(
  image,
  offset,
  _shadowPainter,
);
```

<details>
<summary>Perf with 192 prod objects, no shadow image painters, about same perf between c50b02f2c3 and 02a182dd96</summary>
192 objects, c50b02f2c3 no shadows:
<img width="969" alt="c50b02f2c3_192pieces_noshadow" src="https://user-images.githubusercontent.com/29335974/172062740-2fba0acc-f862-4e50-b24f-345e375977aa.png">

192 objects, 02a182dd96 no shadows (also appears possibly slightly slower overall if forgetting isComplex):
<img width="960" alt="02a182dd96_192pieces_noshadow" src="https://user-images.githubusercontent.com/29335974/172062725-9d2b8fdc-da7a-494e-bb85-89d82cd184fe.png">
</details>

<details>
<summary>Perf with 192 prod objects, with shadow image painters, <=10ms (c50b02f2c3) vs >=60ms (02a182dd96)</summary>
192 objects, c50b02f2c3 with shadows, basically no extra cost over no shadows (assuming complexity was automatically detected):
<img width="964" alt="c50b02f2c3_192pieces_withshadow" src="https://user-images.githubusercontent.com/29335974/172062767-2ab96d8e-a118-4ec6-9aff-e428591495cb.png">

192 objects, 02a182dd96 with shadows, extreme cost increase (but only when not forcing isComplex):
<img width="968" alt="02a182dd96_192pieces_withshadow" src="https://user-images.githubusercontent.com/29335974/172062772-a4e3d2dc-ac82-4c15-9673-0ed6293333ac.png">
</details>

<details>
<summary>Perf with 192 prod objects, shadow paint vs duplicate layer + shadow widget</summary>
192 objects, c50b02f2c3 (with or without isComplex) and 02a182dd966 (with isComplex) with shadow painter, best performance:
<img width="964" alt="c50b02f2c3_192pieces_withshadow" src="https://user-images.githubusercontent.com/29335974/172062767-2ab96d8e-a118-4ec6-9aff-e428591495cb.png">

192 objects, c50b02f2c3 with duplicate object layer (no shadow painter) + shadow widget wrapper, second best performance:
<img width="935" alt="c50b02f2c3_192pieces_duplicateshadowlayer" src="https://user-images.githubusercontent.com/29335974/172062798-4e1bd6c9-a589-4728-a1e8-0d32e82aaf57.png">

192 objects, 02a182dd96 with duplicate object layer (no shadow painter) + shadow widget wrapper, just under 16ms threshold:
<img width="940" alt="02a182dd96_192pieces_duplicateshadowlayer" src="https://user-images.githubusercontent.com/29335974/172062803-05f328e4-6684-489d-88ac-3737cdeec4a8.png">
</details>

The custom painter + picture recorder images work great for the foregrounds, but if one goes to 3.0+ they need to force `isComplex` to true otherwise there will be a dramatic performance drop. With custom painters this is okay, since it is controllable. But I believe other widgets like ImageFiltered/ColorFiltered may need to be aware of this and/or have a similar option to control it. Here is an example between the two commits when adding in some icon widgets with shadows.

<details>
<summary>Perf with 24 icons, ImageFiltered for shadows</summary>
24 icons, c50b02f2c3 with ImageFiltered/ColorFiltered widget combo to add shadow:
<img width="969" alt="c50b02f2c3_24icons_withshadow" src="https://user-images.githubusercontent.com/29335974/172062820-c856b7af-0cd1-4eda-91ef-76616145a70e.png">

24 icons, 02a182dd96 with ImageFiltered/ColorFiltered widget combo to add shadow:
<img width="972" alt="02a182dd96_24icons_withshadow" src="https://user-images.githubusercontent.com/29335974/172062816-b60db995-e6c1-4705-91f6-bc230966e1fd.png">
</details>

Otherwise it appears developers will need to convert from using the pre-created widgets over to making their own CustomerPainters + images, if they need to maximize performance. Less impactful than the initial issue, but can still lead to a buildup of extra costs as more are added to a page.

Thanks again for the tip on the PictureRecorder, and sorry I keep trying to break stuff further 😄
